cmd/libsnap-confine-private/snap.c

```diff
@@ -28,21 +28,26 @@
 #include "string-utils.h"
 #include "cleanup-funcs.h"
 
-bool sc_security_tag_validate(const char *security_tag, const char *snap_name)
+bool sc_security_tag_validate(const char *security_tag,
+			      const char *snap_instance,
+			      const char *component_name)
 {
 	/* Don't even check overly long tags. */
 	if (strlen(security_tag) > SNAP_SECURITY_TAG_MAX_LEN) {
 		return false;
 	}
 	const char *whitelist_re =
-	    "^snap\\.([a-z0-9](-?[a-z0-9])*(_[a-z0-9]{1,10})?)\\.([a-zA-Z0-9](-?[a-zA-Z0-9])*|hook\\.[a-z](-?[a-z0-9])*)$";
+	    "^snap\\.([a-z0-9](-?[a-z0-9])*(_[a-z0-9]{1,10})?)(\\.[a-zA-Z0-9](-?[a-zA-Z0-9])*|(\\+([a-z0-9](-?[a-z0-9])*))?\\.hook\\.[a-z](-?[a-z0-9])*)$";
 	regex_t re;
 	if (regcomp(&re, whitelist_re, REG_EXTENDED) != 0)
 		die("can not compile regex %s", whitelist_re);
 
 	// first capture is for verifying the full security tag, second capture
-	// for verifying the snap_name is correct for this security tag
-	regmatch_t matches[2];
+	// for verifying the snap_name is correct for this security tag, eighth capture
+	// for verifying the component_name is correct for this security tag. the
+    // expression currently contains 9 capture groups, but we only care about
+    // these 3.
+	regmatch_t matches[9];
```

Comment:
btw, I still think we should:

```
if (sizeof matches / sizeof matches[0] != re.re_nsub) {
    die("internal error: regex not fully accounted for");
}
```
hopefully this will blow up if we modify the regex and not account for something.